lib/active_merchant/billing/gateways/hi_pay.rb

```diff
@@ -132,6 +139,37 @@ def tokenize(payment_method, options = {})
         commit('store', post, options)
       end
 
+      def add_3ds(post, options)
+        return unless options.has_key?(:execute_threed)
+
+        browser_info_3ds = options[:three_ds_2][:browser_info]
+
+        browser_info_hash = {
+          "java_enabled": browser_info_3ds[:java],
+            "javascript_enabled": (browser_info_3ds[:javascript] || false),
+            "ipaddr":  options[:ip],
+            "http_accept": '*\\/*',
+            "http_user_agent": browser_info_3ds[:user_agent],
+            "language": browser_info_3ds[:language],
+            "color_depth": browser_info_3ds[:depth],
+            "screen_height":   browser_info_3ds[:height],
+            "screen_width": browser_info_3ds[:width],
+            "timezone": browser_info_3ds[:timezone]
+        }
+
+        browser_info_hash['device_fingerprint'] = options[:device_fingerprint] if options[:device_fingerprint]
+        post[:browser_info] = browser_info_hash.to_json
+        post.to_json
+
+        post[:accept_url] = options[:accept_url]
+        post[:decline_url] = options[:decline_url]
+        post[:pending_url] = options[:pending_url]
+        post[:exception_url] = options[:exception_url]
+        post[:cancel_url] = options[:cancel_url]
+        post[:notify_url] = browser_info_3ds[:notification_url]
```

Comment:
All this params would be null if they are not specified, maybe we can add an if validation for each one.